akka-http-core/src/main/scala/akka/http/engine/client/HostConnectionPoolConductor.scala

```diff
@@ -0,0 +1,230 @@
+               | Preferred)|     |   Merge)  |               |             |     |   Route)  +-------------->
+               +----+------+     +-----+-----+               +-------------+     +-----------+       to slots     
+                    ^                  ^                                                               
+                    |                  | SimpleSlotEvent                                              
+                    | Request-         |                                                               
+                    | Context     +---------+
+                    +-------------+  retry  |<-------- Slot Event (from slotEventMerge)
+                                  |  Split  |
+                                  +---------+
+
+  */
+  def apply(slotCount: Int, maxRetries: Int, pipeliningLimit: Int, log: LoggingAdapter): Graph[Ports, Any] =
+    FlowGraph.partial() { implicit b ⇒
+      import FlowGraph.Implicits._
+
+      val merge1 = b.add(new Merge1)
+      val merge2 = b.add(new Merge2(slotCount, maxRetries, pipeliningLimit, log))
+      val doubler = Flow[SwitchCommand].mapConcat(x ⇒ x :: x :: Nil)
+      val route = b.add(new Route(slotCount))
+      val retrySplit = b.add(new RetrySplit())
+
+      merge1.out ~> merge2.in0
+      merge2.out ~> doubler ~> route.in
+      retrySplit.out0 ~> merge2.in1
+      retrySplit.out1 ~> merge1.in1
+
+      Ports(merge1.in0, retrySplit.in, route.outlets.asInstanceOf[immutable.Seq[Outlet[RequestContext]]])
+    }
+
+  // almost the same as a MergePreferred,
+  // but as MergePreferred doesn't appear to propagate completion on the secondary input we can use it here
+  private class Merge1 extends FlexiMerge[RequestContext, FanInShape2[RequestContext, RequestContext, RequestContext]](
+    new FanInShape2("Merge1"), OperationAttributes.name("Merge1")) {
+    import FlexiMerge._
+
+    def createMergeLogic(shape: FanInShape2[RequestContext, RequestContext, RequestContext]) =
+      new MergeLogic[RequestContext] {
+        val freshInlet = shape.in0
+        val retryInlet = shape.in1
+
+        override def initialState = State(ReadAny(retryInlet, freshInlet)) {
+          case (ctx, `retryInlet`, rc: RequestContext) ⇒ { ctx.emit(rc); SameState }
+          case (ctx, `freshInlet`, rc: RequestContext) ⇒ { ctx.emit(rc); SameState }
+        }
+
+        override def initialCompletionHandling = eagerClose
+      }
+  }
+
+  private case class SwitchCommand(rc: RequestContext, slotIx: Int)
+
+  private sealed trait SlotState
+  private case object Unconnected extends SlotState
+  private case object Idle extends SlotState
+  private final case class Loaded(openIdempotentRequests: Int) extends SlotState { require(openIdempotentRequests > 0) }
+  private case class Busy(openRequests: Int) extends SlotState { require(openRequests > 0) }
+  private object Busy extends Busy(1)
+
+  private class Merge2(slotCount: Int, maxRetries: Int, pipeliningLimit: Int, log: LoggingAdapter)
```

Comment:
Absolutely, good idea!
I'll also add more internal docs for the class and the slot state model.